src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -383,7 +383,60 @@ protected void InitializeReferences()
             RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
             try
             {
-                return runner.Scan(this, input, beginning, beginning + length, startat, prevlen, quick, internalMatchTimeout);
+                runner.InitializeTimeout(internalMatchTimeout);
+                ReadOnlySpan<char> span = input.AsSpan(beginning, length);
+                runner.InitializeForScan(this, span, 0, span.Length, startat - beginning, quick);
+                runner.InitializeForGo();
+                runner.Scan(this, span, startat - beginning, prevlen, quick);
+                Match? m = runner.runmatch;
+                runner.runmatch = null; // Reset runmatch
+                if (m is not null)
+                {
+                    if (!quick)
+                    {
+                        if (m._matchcount[0] > 0)
+                            m.Tidy(runner.runtextpos);
+                        if (m.Text != input)
+                            m.Text = input;
+                    }
+
+                    // If there was a match and the original text was sliced, then add beggining to the index to get the real
+                    // Index of the match.
+                    if (m.Success && beginning != 0)
+                    {
+                        m.AddBeginningToIndex(beginning);
+                    }
+                }
+                return m;
+            }
+            finally
+            {
+                _runner = runner;
+            }
+        }
+
+        internal Match? Run(bool quick, int prevlen, ReadOnlySpan<char> input, int beginning, int length, int startat)
+        {
+            if ((uint)startat > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.startat, ExceptionResource.BeginIndexNotNegative);
+            }
+            if ((uint)length > (uint)input.Length)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.length, ExceptionResource.LengthNotNegative);
+            }
+
+            RegexRunner runner = Interlocked.Exchange(ref _runner, null) ?? CreateRunner();
+            try
+            {
+                runner.InitializeTimeout(internalMatchTimeout);
+                ReadOnlySpan<char> span = input.Slice(beginning, length);
+                runner.InitializeForScan(this, span, 0, span.Length, startat - beginning, quick);
+                runner.InitializeForGo();
+                runner.Scan(this, span, startat - beginning, prevlen, quick);
+                Match? m = runner.runmatch;
+                runner.runmatch = null; // Reset runmatch
```

Comment:
Doesn't this mean we'll never reuse Match objects from call to call, even if quick == true?

Reply:
yeah, I still have to figure out what the best way to do this. Should be fixable, but the issue right now is that when quick is true and you find a match, you actually want to return null as that is what we today interpret as "there was a match" like:

https://github.com/dotnet/runtime/blob/05dbb13d09f4897cfafa35813a3c7fd9f4df5956/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Match.cs#L38

Given Scan doesn't return anything today, and basically uses runmatch as the communication mechanism, then that is why even when quick is true we are setting it to null. That said, after this new refactoring where all of the handling of runmatch happens as part of Run(), this should be easy to fix.

Reply:
I'm not following.  We don't currently null out runmatch for quick; why do we need to here?
https://github.com/dotnet/runtime/blob/05dbb13d09f4897cfafa35813a3c7fd9f4df5956/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs#L179-L183

Reply:
We don't need it currently because the current `Scan` returns a Match object, so when quick == true and we found a match, we simply return null:

https://github.com/dotnet/runtime/blob/87777b624b0b8ea295b0fe9cd8505ce262ed75c0/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs#L175-L183

The new Scan method however returns void, and the way to communicate the result back to Run() used to be via runmatch (before I moved all of the handling of runmatch to inside Run() method), which is why I was setting runmatch to null when quick was true. Now that all of the handling of runmatch happens during Run(), I can avoid nulling out runmatch and simply return null from `Run` method whenever we found a match and quick was set to true.